crates/rome_js_analyze/src/analyzers/complexity/no_confusing_void_type.rs

```diff
@@ -0,0 +1,281 @@
+    /// This option is `false` by default.
+    ///
+    /// The following patterns are considered warnings with `{ allowAsThisParameter: false }` but valid with `{ allowAsThisParameter: true }`:
+    ///
+    /// ```ts
+    /// function doThing(this: void) {}
+    /// class Example {
+    ///     static helper(this: void) {}
+    ///     callback(this: void) {}
+    /// }
+    /// ```
+    ///
+    pub(crate) NoConfusingVoidType {
+        version: "1.0.0",
+        name: "noConfusingVoidType",
+        recommended: false,
+    }
+}
+
+/// rule options for `noConfusingVoidType`
+#[derive(Debug, Clone, Deserialize, Serialize, Bpaf)]
+#[cfg_attr(feature = "schema", derive(JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoConfusingVoidTypeOptions {
+    /// This option lets you control if void can be used as a valid value for generic type parameters.
+    allow_in_generic_type_arguments: bool,
+    /// This option allows specifying a this parameter of a function to be void when set to true. This pattern can be useful to explicitly label function types that do not use a this argument.
+    allow_as_this_parameter: bool,
+}
+
+impl NoConfusingVoidTypeOptions {
+    const KNOWN_KEYS: &'static [&'static str] =
+        &["allowInGenericTypeArguments", "allowAsThisParameter"];
+}
+
+impl Default for NoConfusingVoidTypeOptions {
+    fn default() -> Self {
+        Self {
+            allow_in_generic_type_arguments: true,
+            allow_as_this_parameter: false,
+        }
+    }
+}
+
+type Language = <AnyTsType as AstNode>::Language;
+
+// We only focus on union type
+pub enum VoidTypeIn {
+    Union,
+    Unknown,
+}
+
+impl Rule for NoConfusingVoidType {
+    type Query = Ast<AnyTsType>;
+
+    type State = VoidTypeIn;
+
+    type Signals = Option<Self::State>;
+
+    type Options = NoConfusingVoidTypeOptions;
+
+    fn run(ctx: &rome_analyze::context::RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        if let AnyTsType::TsVoidType(node) = node {
+            let result = node_in(node.syntax(), ctx.options());
+            return result;
+        }
+
+        None
+    }
+    fn diagnostic(
+        ctx: &rome_analyze::context::RuleContext<Self>,
+        state: &Self::State,
+    ) -> Option<rome_analyze::RuleDiagnostic> {
+        let options = ctx.options();
+        let node = ctx.query();
+        return Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.range(),
+            markup! {{match_message(state, options)}},
+        ));
+    }
+}
+
+fn node_in(
+    node: &SyntaxNode<Language>,
+    options: &NoConfusingVoidTypeOptions,
+) -> Option<VoidTypeIn> {
+    let parent = node.parent()?;
+
+    // (string | void)
+    if let Some(n) = parent.cast_ref::<TsParenthesizedType>() {
+        return node_in(n.syntax(), options);
```

Comment:
```suggestion
    if TsParenthesizedType::can_cast(parent.kind()) {
        return node_in(parent, options);
```

You can avoid cloning the node by using `can_cast`.